sway-core/src/asm_generation/fuel/optimizations.rs

```diff
@@ -0,0 +1,257 @@
+                            }
+                        }
+                    }
+                    VirtualOp::LoadDataId(dest, data_id) => {
+                        if let Some(c) = data_section.get_data_word(data_id) {
+                            reg_contents.insert(dest.clone(), RegContents::Constant(c));
+                        } else {
+                            reg_contents.remove(dest);
+                        }
+                        record_new_def(&mut latest_version, dest);
+                    }
+                    VirtualOp::MOVI(dest, imm) => {
+                        reg_contents.insert(dest.clone(), RegContents::Constant(imm.value as u64));
+                        record_new_def(&mut latest_version, dest);
+                    }
+                    VirtualOp::LW(dest, addr_reg, imm) => match reg_contents.get(addr_reg) {
+                        Some(RegContents::BaseOffset(base_reg, offset))
+                            if get_def_version(&latest_version, &base_reg.reg) == base_reg.ver
+                                && ((offset / 8) + imm.value as u64)
+                                    < compiler_constants::TWELVE_BITS =>
+                        {
+                            let new_imm = VirtualImmediate12::new_unchecked(
+                                (offset / 8) + imm.value as u64,
+                                "Immediate offset too big for LW",
+                            );
+                            let new_lw = VirtualOp::LW(dest.clone(), base_reg.reg.clone(), new_imm);
+                            // The register defined is no more useful for us. Forget anything from its past.
+                            reg_contents.remove(dest);
+                            record_new_def(&mut latest_version, dest);
+                            // Replace the LW with a new one in-place.
+                            *op = new_lw;
+                        }
+                        _ => {
+                            reg_contents.remove(dest);
+                            record_new_def(&mut latest_version, dest);
+                        }
+                    },
+                    VirtualOp::SW(addr_reg, src, imm) => match reg_contents.get(addr_reg) {
+                        Some(RegContents::BaseOffset(base_reg, offset))
+                            if get_def_version(&latest_version, &base_reg.reg) == base_reg.ver
+                                && ((offset / 8) + imm.value as u64)
+                                    < compiler_constants::TWELVE_BITS =>
+                        {
+                            let new_imm = VirtualImmediate12::new_unchecked(
+                                (offset / 8) + imm.value as u64,
+                                "Immediate offset too big for SW",
+                            );
+                            let new_sw = VirtualOp::SW(base_reg.reg.clone(), src.clone(), new_imm);
+                            // Replace the SW with a new one in-place.
```

Comment:
Hmmm, why are we not removing the `dest` from ``reg_contents` in the case of `SW`? And also recording the new definition of `dest` like in `LW`?

Reply:
`SW` doesn't define any register. It just stores the value in one register to the memory location pointed to by the other.